### ISSUE TITLE
Connection not released to Agroal connection pool when JPA query using getResultStream runs outside transaction

### ISSUE THREAD

**Issue by mrickly** [quarkusio/quarkus#36643]
### Describe the bug

Given a `jakarta.persistence.TypedQuery` created by a `jakarta.persistence.EntityManager` in a Quarkus application, executed outside of a transaction and using `getResultStream()` to access the result set, then the Agroal connection used by the statement is not released back to the pool at the end of the statement. 

### Expected behavior

The Agroal connection used to execute the query should be returned to the pool after execution of the statement in any case. 

### Actual behavior

If `query.getResultStream()` is used to access the result set, then `logicalConnectionManagedImpl.afterStatement()` is called. Since the `PhysicalConnectionHandlingMode` is `DELAYED_ACQUISITION_AND_RELEASE_BEFORE_TRANSACTION_COMPLETION`, the connection is not released. The `logicalConnectionManagedImpl.beforeTransactionCompletion()` and `logicalConnectionManagedImpl.afterTransaction()` methods that would release the connection given `DELAYED_ACQUISITION_AND_RELEASE_BEFORE_TRANSACTION_COMPLETION` are never called. 

If `query.getSingleResult()` is used to access the result set, then the `logicalConnectionManagedImpl.afterTransaction()` method is called and it releases the connection to the pool. 

### How to Reproduce?

```
    @Transactional(Transactional.TxType.NOT_SUPPORTED)
    public Product getProductByProductId(String productId) {
        TypedQuery<Product> query = entityManager.createQuery("select p from Product p where p.productId = :productId", Product.class);
        query.setParameter("productId", productId);
        Stream<Product> productStream = query.getResultStream();
        Optional<Product> productOptional = productStream.findFirst();
        productStream.close();
        Product product = productOptional.get();
        return product;
    }
```
*Note that the issue cannot be reproduced if this method is called from a REST resource*, because then the `destroy` method of the `RequestScopedSessionHolder` will release the connection, but it can be reproduced when the above code is called from a messaging `@Incoming` method. 

### Additional information

Opening this issue as a follow up to the discussion with @yrodiere in https://github.com/quarkusio/quarkus/issues/28619. 
The behaviour is documented for [Panache](https://quarkus.io/guides/hibernate-orm-panache#most-useful-operations), but not for [JPA/Hibernate](https://quarkus.io/guides/hibernate-orm) as far as I can tell.

**Comment by yrodiere** [comment]
Hey,

Thanks for reporting.

> Note that the issue cannot be reproduced if this method is called from a REST resource

That's a piece of information I was missing. So as long as you're in a transaction OR a request scope is active, everything works fine. It's only faulty when using `@Transactional(Transactional.TxType.NOT_SUPPORTED)`, which is indeed a problem since this is the recommended way of disabling transactions.

> but it can be reproduced when the above code is called from a messaging @Incoming method.

Are we talking of a *reactive* messaging method? You did annotate the method with `@Blocking`, right? Otherwise using Hibernate ORM in a reactive context is invalid, and in fact I would have expected this to fail, because Vert.x really doesn't like you blocking the event loop and there are some checks in place to prevent that. 

> If query.getSingleResult() is used to access the result set, then the logicalConnectionManagedImpl.afterTransaction() method is called and it releases the connection to the pool.

I'm surprised this doesn't happen on `Stream#close`... Needs some investigation.

**Comment by mrickly** [comment]
@yrodiere : Correct, we are using `@Blocking` on the `@Incoming` method. With that setup the `@Transactional(Transactional.TxType.NOT_SUPPORTED)` is actually superfluous.

**Comment by gsmet** [comment]
I was going to have a look at this one and check if it was still a problem but it would help to have a reproducer (sorry, I know it's late...).

Any chance you could provide a simple reproducer so that we can have a look?

**Comment by mrickly** [comment]
@gsmet : Thanks for looking into this. On our end, we told developers to make sure to run within a transaction when calling query.getResultStream(). Nevertheless, it seems that the issue described above still exists and I can provide a small reproducer.

There are two tests: 

- The first one (by Order) shows that the unique connection in the jdbc thread pool is released when the bean method which contains `query.getResultStream()` and which is annotated with `@Transactional(Transactional.TxType.NOT_SUPPORTED)` is called via a REST endpoint. 
- The second one shows that the connection is not released when the bean method is called directly. 

[quarkus-issue-36643-reproducer.zip](https://github.com/user-attachments/files/18270569/quarkus-issue-36643-reproducer.zip)
